### PR TITLE
CE-94: Adding in the Crossref relationship types

### PR DESCRIPTION
**Pull request by ialarmedalien**
Adding in Crossref relationship types

### CODE REVIEW

File: src/us/kbase/workspace/database/provenance/RelationshipType.java [hunk 1]
```diff
@@ -2,72 +2,166 @@
...
+
 
-	private final static String SOURCE = "DataCite";
 	private final String identifier;
+	private final String source;
+
+	// From the DataCite / Crossref "related identifier" crosswalk
+	// The following are Crossref relationship type terms that duplicate
+	// DataCite terms. If any of the following terms is entered,
+	// it should be replaced by the DataCite term.
```
Comment (ialarmedalien) [inline src/us/kbase/workspace/database/provenance/RelationshipType.java:88]: group decision to use DataCite as the primary relationship type supplier.

File: src/us/kbase/workspace/database/provenance/RelationshipType.java [hunk 2]
```diff
@@ -2,72 +2,166 @@
...
+			// isoriginalformof
+			STRING_TO_REL_TYPE_MAP.put(rt.getIdentifier().toLowerCase(), rt);
+			// datacite:is_original_form_of
+			STRING_TO_REL_TYPE_MAP.put(rt.getSource().toLowerCase() + ":" + rt.name().toLowerCase(), rt);
+			// datacite:isoriginalformof
 			STRING_TO_REL_TYPE_MAP.put(rt.getPid().toLowerCase(), rt);
+			// add in the crossref equivalent (if applicable)
+			if (CROSSREF_DUPLICATES.contains(rt.getIdentifier())) {
+				// e.g. crossref:is_variant_form_of
+				STRING_TO_REL_TYPE_MAP.put("crossref:" + rt.name().toLowerCase(), rt);
```
Comment (MrCreosote) [inline src/us/kbase/workspace/database/provenance/RelationshipType.java:127]: I would make constants for `DataCite`, `Crossref`, and `crossref:`

File: src/us/kbase/workspace/database/provenance/RelationshipType.java [hunk 3]
```diff
@@ -2,72 +2,166 @@
...
+	}
+
+	/**
+	 * Get the identifier for this relationship type.
+	 *
+	 * @return the identifier.
+	 */
+	private String getIdentifier() {
+		return identifier;
+	}
```
Comment (MrCreosote) [inline src/us/kbase/workspace/database/provenance/RelationshipType.java:155]: There's no real need for these methods I think - you can just access via `rt.source` or `rt.identifier` IIRC

File: src/us/kbase/workspace/database/provenance/RelationshipType.java [hunk 4]
```diff
@@ -2,72 +2,166 @@
...
-	private final static String SOURCE = "DataCite";
 	private final String identifier;
+	private final String source;
+
+	// From the DataCite / Crossref "related identifier" crosswalk
+	// The following are Crossref relationship type terms that duplicate
+	// DataCite terms. If any of the following terms is entered,
+	// it should be replaced by the DataCite term.
+	private static final List<String> CROSSREF_DUPLICATES = Arrays.asList(
+		"Compiles",
```
Comment (MrCreosote) [inline src/us/kbase/workspace/database/provenance/RelationshipType.java:90]: As these strings are used in 2 places you might consider constants for them

File: src/us/kbase/workspace/test/database/provenance/RelationshipTypeTest.java [hunk 5]
```diff
@@ -23,24 +23,30 @@ public void testGetPid() throws Exception {
 		assertThat("incorrect PID",
 				rt2.getPid(),
 				is("DataCite:IsPreviousVersionOf"));
+		final RelationshipType rt3 = RelationshipType.IS_MANIFESTATION_OF;
+		assertThat("incorrect PID",
+				rt3.getPid(),
+				is("Crossref:IsManifestationOf"));
 	}
 
 	@Test
```
Comment (MrCreosote) [inline src/us/kbase/workspace/test/database/provenance/RelationshipTypeTest.java:32]: It'd be good to have a test that goes from a string -> non-duplicate crossref type

File: src/us/kbase/workspace/database/provenance/RelationshipType.java (hunk 4 above)
Comment (MrCreosote) [reply src/us/kbase/workspace/database/provenance/RelationshipType.java:91]: FINE, geez

File: src/us/kbase/workspace/test/database/provenance/RelationshipTypeTest.java (hunk 5 above)
Comment (MrCreosote) [reply src/us/kbase/workspace/test/database/provenance/RelationshipTypeTest.java:38]: I meant the other way, using the string -> enum map

Comment (MrCreosote) [reply src/us/kbase/workspace/test/database/provenance/RelationshipTypeTest.java:38]: Not that this is bad

File: src/us/kbase/workspace/database/provenance/RelationshipType.java (hunk 2 above)
Comment (MrCreosote) [reply src/us/kbase/workspace/database/provenance/RelationshipType.java]: :+1:

File: src/us/kbase/workspace/database/provenance/RelationshipType.java (hunk 3 above)
Comment (MrCreosote) [reply src/us/kbase/workspace/database/provenance/RelationshipType.java]: :+1:

File: src/us/kbase/workspace/test/database/provenance/RelationshipTypeTest.java (hunk 5 above)
Comment (MrCreosote) [reply src/us/kbase/workspace/test/database/provenance/RelationshipTypeTest.java:38]: :+1:

**Review by MrCreosote — APPROVED** [review]
LGTM

**Pull request merged**